llvm/include/llvm/ABI/Types.h

```diff
@@ -0,0 +1,427 @@
+
+  bool isUnion() const { return Flags.IsUnion; }
+  bool isCXXRecord() const { return Flags.IsCXXRecord; }
+  bool isPolymorphic() const { return Flags.IsPolymorphic; }
+  bool canPassInRegisters() const { return Flags.CanPassInRegisters; }
+  bool hasFlexibleArrayMember() const { return Flags.HasFlexibleArrayMember; }
+
+  uint32_t getNumBaseClasses() const { return BaseClasses.size(); }
+  uint32_t getNumVirtualBaseClasses() const {
+    return VirtualBaseClasses.size();
+  }
+  bool isTransparentUnion() const { return Flags.IsTransparent; }
+  ArrayRef<FieldInfo> getFields() const { return Fields; }
+  ArrayRef<FieldInfo> getBaseClasses() const { return BaseClasses; }
+  ArrayRef<FieldInfo> getVirtualBaseClasses() const {
+    return VirtualBaseClasses;
+  }
+};
+
+/// TypeBuilder manages the lifecycle of ABI types using bump pointer
+/// allocation. Types created by a TypeBuilder are valid for the lifetime of the
+/// builder.
+///
+/// Example usage:
+/// \code
+///   BumpPtrAllocator Alloc;
+///   TypeBuilder Builder(Alloc);
+///   const auto *IntTy = Builder.getIntegerType(32, Align(4), true);
+/// \endcode
+class TypeBuilder {
+private:
+  BumpPtrAllocator &Allocator;
+
+public:
+  explicit TypeBuilder(BumpPtrAllocator &Alloc) : Allocator(Alloc) {}
+
+  const VoidType *getVoidType() {
+    return new (Allocator.Allocate<VoidType>()) VoidType();
+  }
+
+  const IntegerType *getIntegerType(uint64_t BitWidth, Align Align, bool Signed,
+                                    bool IsBitInt = false) {
+    return new (Allocator.Allocate<IntegerType>())
+        IntegerType(BitWidth, Align, Signed, IsBitInt);
+  }
+
+  const FloatType *getFloatType(const fltSemantics &Semantics, Align Align) {
+    return new (Allocator.Allocate<FloatType>()) FloatType(Semantics, Align);
+  }
+
+  const PointerType *getPointerType(uint64_t Size, Align Align,
+                                    unsigned Addrspace = 0) {
+    return new (Allocator.Allocate<PointerType>())
+        PointerType(Size, Align, Addrspace);
+  }
+
+  const ArrayType *getArrayType(const Type *ElementType, uint64_t NumElements,
+                                uint64_t SizeInBits,
+                                bool IsMatrixType = false) {
+    return new (Allocator.Allocate<ArrayType>())
+        ArrayType(ElementType, NumElements, SizeInBits, IsMatrixType);
+  }
+
+  const VectorType *getVectorType(const Type *ElementType,
+                                  ElementCount NumElements, Align Align) {
+    return new (Allocator.Allocate<VectorType>())
+        VectorType(ElementType, NumElements, Align);
+  }
+
+  const RecordType *
+  getRecordType(ArrayRef<FieldInfo> Fields, TypeSize Size, Align Align,
+                StructPacking Pack = StructPacking::Default,
+                ArrayRef<FieldInfo> BaseClasses = {},
+                ArrayRef<FieldInfo> VirtualBaseClasses = {},
+                bool CXXRecord = false, bool Polymorphic = false,
```

Comment:
Sequences of default-value boolean arguments are bad for readability, and are hard to refactor. For a widely-used API, we want to have something that's easy to update. For example, to make this handle MSVC ABI lowering, we're going to need some kind of "non-trivially copyable" bit. I recommend using a flag enum instead, where the call site looks like:

  thing().getRecordType(Fields, Size, Align, Packing, Bases, VBases, RecordFlags::IsCxxRecord | RecordFlags::IsPolymorphic | RecordFlags::CanPassInRegisters | RecordFlags::FlexibleArray | RecordFlags::NonTrivial)

Or you build up the flags with `|=` statements at the call site. This makes it possible to remove flags in the future without changing the meaning of sequences of `, false, true, false)`.